### PR TITLE
Optimizer for fine tuning, Optional support of TensorboardX

### PR DESCRIPTION
**Pull request by SunQpark**
Hi, this PR is 3 small fixes on your work.

First, I changed the optimizer call in the base_trainer to make it work with fine-tuning, where models have non-trainable parameters.

Second is minor fix on total number of train data displayed, which is changing 
Train Epoch: 1 [5120/**53984** (9%)] Loss: 0.912355
to
Train Epoch: 1 [5120/**54000** (9%)] Loss: 0.912355.

Last one is adding items to .gitignore file, most of which are from default .gitignore file of github for python.

### CODE REVIEW

**Comment by SunQpark** [comment]
Here are explanation for the last 6 commits.

I replaced trainer.gpu with trainer.device, which points to cpu if gpu is not available or cuda option is turned off. I deprecated trainer._to_tensor() since casting data and target to float and long is specific only to classification task, and even if we need that, dataset or data_loader should be responsible for that.

Modified _evaluate_metrics() not to take argmax on the output since it is also classification specific.
Changed example metric functions to accuracy and top-k accuracy, which are written with pure pytorch.

I added tensorboardX visualization in the logger/visualization.py. My implementation of this class is supposed to be harmless when tensorboardX is not installed or it is turned off in config file, but I did not tested this on a machine without tensorboardX.

**Comment by borgesa** [comment]
@SunQpark ,

Well done. It runs well on my laptop (CPU installation of Tensorflow).

**In README.md**: Maybe you should explicitly mention that Tensorflow needs to be installed as well?

@victoresque: Will you merge this request into the repository? Or do you have comments/concerns?

**Comment by victoresque** [comment]
Hi @borgesa, I will merge this PR and then update the requirements in `README.md`, 

@SunQpark thanks for the contribution, the implementation works well

**Comment by SunQpark** [comment]
Hi, @borgesa 
Indeed, I already considered using `config['visualization']['type'] = 'tensorboardX'`. 
I did not used it that way because I thought it would be not comfortable to turn on/off.
Consider some people who wants to turn off the visualization on the config file. 
If it reads `config['visualization']['type'] = 'tensorboardX'`, people may try `config['visualization']['type'] = ''` or just delete the line. Both way works fine when turning off visualization, but typing the 'tensorboardX' or the whole line of config when turning the visualization on again would be an annoying task to do(what if some typo occurs in the string 'tensorboardX', or don't remember what the line looked like before deleting?). I will frequently toggle the visualization myself because I don't want the log to be saved while I'm debugging my code, so I choose this way. I also wants to say that in this way we can use two visualizations at the same time, but that will be too much...

However, I agree that it would be nice if we can support both visdom and tensorboard with same writer instance. From now on, I think I can handle that problem with inheriting `__getattribute__` method but I have another problem that I don't know much about visdom. I will consider making another PR to generalize visualization and will help If there are someone tries to add visdom.
Thanks for your suggestion.

**Comment by borgesa** [comment]
Hi,

Ok, I definitely agree with the decision in the configuration file.

As I understand you, I believe we agree on the longer time goal (that for instance will lead to 
'base_trainer' being generalized to get appropriate writer[s]).

**Comment by SunQpark** [comment]
Yes, I agree.
Tell me if you think there is anything I can help
